### PR TITLE
Check for existance of "adaptiveFormats" in videos

### LINKED ISSUE #2171: [Bug] Missing hash key: "adaptiveFormats" (KeyError)
**Describe the bug**

I can’t play this video https://redirect.invidious.io/watch?v=RGU-4iVs7X4

**Steps to Reproduce**
Try to play this video https://redirect.invidious.io/watch?v=RGU-4iVs7X4

**Logs**

Title: `Missing hash key: "adaptiveFormats" (KeyError)`
Date: `2021-06-18T12:31:45Z`
Route: `/watch?v=RGU-4iVs7X4`
Version: `2021.06.14-3146678 @ master`
<details>
<summary>Backtrace</summary>
<p>
   
```
Missing hash key: "adaptiveFormats" (KeyError)
  from /usr/share/crystal/src/hash.cr:1027:11 in '[]'
  from /usr/share/crystal/src/json/any.cr:102:7 in '[]'
  from src/invidious/videos.cr:765:34 in 'is_vr'
  from src/invidious/views/watch.ecr:61:13 in 'handle'
  from src/invidious.cr:282:1 in '->'
  from /usr/share/crystal/src/primitives.cr:255:3 in '->'
  from /usr/share/crystal/src/primitives.cr:255:3 in 'process_request'
  from lib/kemal/src/kemal/route_handler.cr:17:7 in 'call'
  from /usr/share/crystal/src/http/server/handler.cr:28:7 in 'call_next'
  from lib/kemal/src/kemal/websocket_handler.cr:13:14 in 'call'
  from /usr/share/crystal/src/http/server/handler.cr:28:7 in 'call_next'
  from lib/kemal/src/kemal/filter_handler.cr:21:7 in 'call'
  from /usr/share/crystal/src/http/server/handler.cr:28:7 in 'call_next'
  from src/invidious/helpers/handlers.cr:212:5 in 'call'
  from /usr/share/crystal/src/http/server/handler.cr:28:7 in 'call_next'
  from src/invidious/helpers/handlers.cr:94:12 in 'call'
  from /usr/share/crystal/src/http/server/handler.cr:28:7 in 'call_next'
  from src/invidious/helpers/handlers.cr:145:12 in 'call'
  from /usr/share/crystal/src/http/server/handler.cr:28:7 in 'call_next'
  from src/invidious/helpers/handlers.cr:70:5 in 'call'
  from /usr/share/crystal/src/http/server/handler.cr:28:7 in 'call_next'
  from src/invidious/helpers/static_file_handler.cr:189:11 in 'call'
  from /usr/share/crystal/src/http/server/handler.cr:28:7 in 'call_next'
  from lib/kemal/src/kemal/exception_handler.cr:8:7 in 'call'
  from /usr/share/crystal/src/http/server/handler.cr:28:7 in 'call_next'
  from src/invidious/helpers/logger.cr:17:35 in 'call'
  from /usr/share/crystal/src/http/server/handler.cr:28:7 in 'call_next'
  from lib/kemal/src/kemal/init_handler.cr:12:7 in 'call'
  from /usr/share/crystal/src/http/server/request_processor.cr:51:11 in 'process'
  from /usr/share/crystal/src/http/server.cr:513:5 in 'handle_client'
  from /usr/share/crystal/src/http/server.cr:468:13 in '->'
  from /usr/share/crystal/src/primitives.cr:255:3 in 'run'
  from /usr/share/crystal/src/fiber.cr:92:34 in '->'
  from ???
```
</p>
</details>

**Screenshots**

**Additional context**

NewPipe only provides a 360p stream, so does youtube-dl. Usually there are 360p and 720p streams that contain both video and audio, called `formatStreams` in the invidious API. `adaptiveFormats` either provide only audio or only video.

The Invidious API returns `[]` so it should be an easy empty check.

### PR DESCRIPTION
**Pull request by syeopite**
Closes #2171

### CODE REVIEW

File: src/invidious/videos.cr [hunk 1]
```diff
@@ -766,7 +766,7 @@ struct Video
   end
 
   def is_vr : Bool
-    info["streamingData"]?.try &.["adaptiveFormats"].as_a[0]?.try &.["projectionType"].as_s == "MESH" ? true : false || false
+    info.dig?("streamingData", "adaptiveFormats", 0, "projectionType").try &.as_s == "MESH" ? true : false || false
```
Comment (SamantazFox) [inline src/invidious/videos.cr:769]: While you're at it, you can refactor this to not use the ternary!

**Review by SamantazFox — APPROVED** [review]

File: src/invidious/videos.cr [hunk 2]
```diff
@@ -765,8 +765,8 @@ struct Video
     info["microformat"]?.try &.["playerMicroformatRenderer"]["isFamilySafe"]?.try &.as_bool || false
   end
 
-  def is_vr : Bool
-    info["streamingData"]?.try &.["adaptiveFormats"].as_a[0]?.try &.["projectionType"].as_s == "MESH" ? true : false || false
+  def is_vr : Bool?
+    return true if info.dig?("streamingData", "adaptiveFormats", 0, "projectionType").try &.as_s == "MESH"
```
Comment (SamantazFox) [inline src/invidious/videos.cr:769]: you should return directly the `==` comparison. if you do `return true if condition`, that will return nil in all cases where the condition is false. You can simply do that:
```
return (info.dig?(...).try &.as_s == "MESH")
```

`info.dig?(...).try &.as_s` will be either a `String` or `nil`, if the element was not found, and then `==` will alway return a `Boolean`.

**Review by SamantazFox — CHANGES_REQUESTED** [review]

File: src/invidious/videos.cr [hunk 3]
```diff
@@ -190,7 +190,7 @@ VIDEO_FORMATS = {
   "248" => {"ext" => "webm", "height" => 1080, "format" => "DASH video", "vcodec" => "vp9"},
   "271" => {"ext" => "webm", "height" => 1440, "format" => "DASH video", "vcodec" => "vp9"},
   # itag 272 videos are either 3840x2160 (e.g. RtoitU2A-3E) or 7680x4320 (sLprVF6d7Ug)
-  "272" => {"ext" => "webm", "height" => 2160, "format" => "DASH video", "vcodec" => "vp9"},
```
Comment (unixfox) [inline src/invidious/videos.cr:193]: Why removing the height?

Comment (syeopite) [reply src/invidious/videos.cr:193]: Oops, that was a remnant of another fix I tried to overload into this PR.

**Pull request merged**